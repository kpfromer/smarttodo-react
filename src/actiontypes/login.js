export const INVALIDATE_LOGIN = 'login/INVALIDATE_LOGIN';
export const REQUEST_LOGIN = 'login/REQUEST_LOGIN';
export const RECEIVE_LOGIN = 'login/RECEIVE_LOGIN';
export const FAILURE_LOGIN = 'login/FAILURE_LOGIN';