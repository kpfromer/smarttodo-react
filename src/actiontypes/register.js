
export const REQUEST_REGISTER = 'register/REQUEST_REGISTER';
export const RECEIVE_REGISTER = 'register/RECEIVE_REGISTER';
export const FAILURE_REGISTER = 'register/FAILURE_REGISTER';