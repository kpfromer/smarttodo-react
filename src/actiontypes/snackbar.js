export const CLOSE_SNACKBAR = 'snackbar/CLOSE_SNACKBAR';
export const CHANGE_SNACKBAR = 'snackbar/CHANGE_SNACKBAR';