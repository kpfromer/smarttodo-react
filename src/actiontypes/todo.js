export const REQUEST_TODOS = 'todo/REQUEST_TODOS';
export const REQUEST_FAILURE_TODOS = 'todo/REQUEST_FAILURE_TODOS';
export const RECIEVE_TODOS = 'todo/RECIEVE_TODOS';

export const ADD_TODO = 'todo/ADD_TODO';
export const ADD_TODO_FAILURE = 'todo/ADD_TODO_FAILURE';
export const ADD_TODO_SUCCESS = 'todo/ADD_TODO_SUCCESS';

export const UPDATE_TODO = 'todo/UPDATE_TODO';
export const UPDATE_TODO_SUCCESS = 'todo/UPDATE_TODO_SUCCESS';
export const UPDATE_TODO_FAILURE = 'todo/UPDATE_TODO_FAILURE';

export const REMOVE_TODO = 'todo/REMOVE_TODO';
export const REMOVE_TODO_SUCCESS = 'todo/REMOVE_TODO_SUCCESS';
export const REMOVE_TODO_FAILURE = 'todo/REMOVE_TODO_FAILURE';