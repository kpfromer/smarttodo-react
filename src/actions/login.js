import * as LoginActionTypes from '../actiontypes/login';
import { CALL_API } from "../middleware/api/api";

export const tryLogin = (username, password) => ({
  [CALL_API]: {
    types: [
      LoginActionTypes.REQUEST_LOGIN,
      LoginActionTypes.RECEIVE_LOGIN,
      LoginActionTypes.FAILURE_LOGIN
    ],
    method: 'POST',
    endpoint: '/login',
    data: { username, password },
    authenticate: false,
    mapResponse: response => response.access_token
  }
});

export const logout = () => ({
  type: LoginActionTypes.INVALIDATE_LOGIN
});