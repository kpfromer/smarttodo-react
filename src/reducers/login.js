import * as LoginActionTypes from '../actiontypes/login';

function login
(
  state = {
    isFetching: false,
    didInvalidate: false,
    token: localStorage.getItem('id') || null
  },
  action
) {
  switch (action.type) {
    case LoginActionTypes.INVALIDATE_LOGIN:
      localStorage.removeItem('id');
      return {
        ...state,
        token: null,
        didInvalidate: true
      };
    case LoginActionTypes.REQUEST_LOGIN:
      return {
        ...state,
        isFetching: true,
        didInvalidate: false
      };
    case LoginActionTypes.RECEIVE_LOGIN:
      localStorage.setItem('id', action.response);
      return {
        ...state,
        isFetching: false,
        didInvalidate: false,
        token: action.response
      };
    case LoginActionTypes.FAILURE_LOGIN:
      return {
        ...state,
        isFetching: false
      };
    default:
      return state;
  }
}

export default login