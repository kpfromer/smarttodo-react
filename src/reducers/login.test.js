import LoginReducer from './login';
import * as LoginActionTypes from "../actiontypes/login";

describe('login reducer', () => {
  let mockGetItem, mockSetItem, mockRemoveItem;

  beforeEach(() => {
    mockGetItem = jest.fn();
    mockSetItem = jest.fn();
    mockRemoveItem = jest.fn();
    global.localStorage = {
      getItem: mockGetItem,
      setItem: mockSetItem,
      removeItem: mockRemoveItem
    };
  });

  it('defaults to not fetching', () => {
    expect(LoginReducer(undefined, 'nsot an action')).toMatchSnapshot();
  });

  it('by default it gets the id from local storage', () => {
    const mockToken = 'jwt-token-in-local-storage';
    mockGetItem.mockReturnValue(mockToken);
    expect(LoginReducer(undefined, 'not an action').token).toBe(mockToken);
    expect(mockGetItem).toHaveBeenCalledWith('id');
  });

  describe('INVALIDATE_LOGIN', () => {
    let result;

    beforeEach(() => {
      result = LoginReducer(undefined, {
        type: LoginActionTypes.INVALIDATE_LOGIN
      });
    });

    it('removes id from local storage', () => {
      expect(mockRemoveItem).toHaveBeenCalledWith('id');
    });

    it('removes token', () => {
      expect(result).toMatchSnapshot();
    });
  });

  describe('RECEIVE_LOGIN', () => {
    let result, token;

    beforeEach(() => {
      token = 'mock-jwt-token-from-api';
      result = LoginReducer(undefined, {
        type: LoginActionTypes.RECEIVE_LOGIN,
        response: token
      });
    });

    it('sets id in local storage', () => {
      expect(mockSetItem).toHaveBeenCalledWith('id', token);
    });

    it('sets token in state ', () => {
      expect(result).toMatchSnapshot();
    });
  });

  describe('REQUEST_LOGIN', () => {
    it('displays spinner', () => {
      const result = LoginReducer(undefined, {
        type: LoginActionTypes.REQUEST_LOGIN
      });
      expect(result).toMatchSnapshot();
    });
  });
});