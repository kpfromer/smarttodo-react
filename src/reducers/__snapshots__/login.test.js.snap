// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login reducer INVALIDATE_LOGIN removes token 1`] = `
Object {
  "didInvalidate": true,
  "isFetching": false,
  "token": null,
}
`;

exports[`login reducer RECEIVE_LOGIN sets token in state  1`] = `
Object {
  "didInvalidate": false,
  "isFetching": false,
  "token": "mock-jwt-token-from-api",
}
`;

exports[`login reducer REQUEST_LOGIN displays spinner 1`] = `
Object {
  "didInvalidate": false,
  "isFetching": true,
  "token": "jwt-token-in-local-storage",
}
`;

exports[`login reducer defaults to not fetching 1`] = `
Object {
  "didInvalidate": false,
  "isFetching": false,
  "token": null,
}
`;
